### PR TITLE
Appsflyer as a source

### PR DESCRIPTION
**Pull request by prabrishac**
## Description of the change

> Description here

## Type of change
- [ ] Bug fix (non-breaking change that fixes an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Related issues

> Fix [#1]() 

### Development

- [ ] Lint rules pass locally
- [ ] The code changed/added as part of this pull request has been covered with tests
- [ ] All tests related to the changed code pass in development

### Code review 

- [ ]  This pull request has a descriptive title and information useful to a reviewer. There may be a screenshot or screencast attached
- [ ] "Ready for review" label attached to the PR and reviewers mentioned in a comment
- [ ] Changes have been reviewed by at least one other engineer
- [ ] Issue from task tracker has a link to this pull request

### CODE REVIEW

File: __tests__/data/appsflyer_source_output.json [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+      }
+    },
+    "integrations": {
+      "Appsflyer": false
+    },
+    "properties": {
+      "idfv": "868049A3-4B2F-4A11-9B00-CFFC362XXXXX",
+      "device_category": "phone",
+      "af_sub1": null,
+      "customer_user_id": "hi.from@appsflyer.example.com",
```
Comment (arnab-p) [inline __tests__/data/appsflyer_source_output.json:19]: Expected: `message.userId`

File: __tests__/data/appsflyer_source_output.json [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+      "is_lat": null,
+      "contributor_2_af_prt": null,
+      "bundle_id": "com.appsflyer.AppsFlyer",
```
Comment (arnab-p) [inline __tests__/data/appsflyer_source_output.json:22]: Expected: `message.context.app.namespace`

File: __tests__/data/appsflyer_source_output.json [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+      "gp_broadcast_referrer": "",
+      "contributor_2_touch_time": null,
+      "contributor_3_touch_type": null,
+      "event_source": "SDK",
+      "af_cost_value": null,
+      "contributor_1_match_type": null,
+      "app_version": "1.4.1",
```
Comment (arnab-p) [inline __tests__/data/appsflyer_source_output.json:29]: Expected: `message.context.app.version`

File: __tests__/data/appsflyer_source_output.json [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+      "contributor_3_af_prt": null,
+      "custom_data": null,
+      "contributor_2_touch_type": null,
+      "gp_install_begin": null,
+      "city": "Khu Pho Binh Hoa",
```
Comment (arnab-p) [inline __tests__/data/appsflyer_source_output.json:34]: Expected: `message.traits.address.city`

File: __tests__/data/appsflyer_source_output.json [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+      "amazon_aid": null,
+      "gp_referrer": null,
+      "af_cost_model": null,
+      "af_c_id": null,
+      "attributed_touch_time_selected_timezone": null,
+      "selected_currency": "USD",
+      "app_name": "AppsFlyer",
```
Comment (arnab-p) [inline __tests__/data/appsflyer_source_output.json:41]: Expected: `message.context.app.name`

File: __tests__/data/appsflyer_source_output.json [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+      "install_time_selected_timezone": "2019-01-20 04:51:16.000+0000",
+      "postal_code": "823941",
```
Comment (arnab-p) [inline __tests__/data/appsflyer_source_output.json:43]: Expected: `message.traits.address.zip`

File: __tests__/data/appsflyer_source_output.json [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+      "wifi": true,
```
Comment (arnab-p) [inline __tests__/data/appsflyer_source_output.json:44]: Expected: `message.context.network.wifi`

File: __tests__/data/appsflyer_source_output.json [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+      "af_adset_id": null,
+      "device_download_time_selected_timezone": "2019-01-20 04:51:16.000+0000",
+      "contributor_2_media_source": null,
+      "conversion_type": null,
+      "contributor_2_match_type": null,
+      "api_version": "2.0",
+      "attributed_touch_time": null,
+      "revenue_in_selected_currency": null,
+      "is_retargeting": false,
+      "country_code": "VN",
```
Comment (arnab-p) [inline __tests__/data/appsflyer_source_output.json:60]: Expected: `message.traits.address.country`

File: __tests__/data/appsflyer_source_output.json [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+      "gp_click_time": null,
+      "contributor_1_af_prt": null,
+      "match_type": null,
+      "appsflyer_id": "1547985076649-5309999",
```
Comment (arnab-p) [inline __tests__/data/appsflyer_source_output.json:64]: Expected

```
context.externalId: [
{
type: "appsflyerExternalId",
value: <value here>
}
]
```

File: __tests__/data/appsflyer_source_output.json [hunk 10]
```diff
@@ -0,0 +1,141 @@
...
+      "store_reinstall": null,
+      "install_app_store": null,
+      "media_source": "organic",
+      "deeplink_url": null,
+      "campaign": null,
+      "af_keywords": null,
+      "region": "AS",
+      "cost_in_selected_currency": null,
+      "event_value": "{}",
+      "ip": "1.1.1.1",
```
Comment (arnab-p) [inline __tests__/data/appsflyer_source_output.json:79]: Expected: `message.context.ip`

File: __tests__/data/appsflyer_source_output.json [hunk 11]
```diff
@@ -0,0 +1,141 @@
...
+      "oaid": null,
+      "event_time": "2020-01-15 14:57:24.898",
```
Comment (arnab-p) [inline __tests__/data/appsflyer_source_output.json:81]: Use it as `timestamp`?

File: __tests__/data/appsflyer_source_output.json [hunk 12]
```diff
@@ -0,0 +1,141 @@
...
+      "af_sub_siteid": null,
+      "advertising_id": null,
+      "os_version": "12.3.1",
+      "platform": "ios",
+      "af_sub3": null,
+      "contributor_3_match_type": null,
+      "selected_timezone": "UTC",
+      "af_ad_id": null,
+      "contributor_3_touch_time": null,
+      "user_agent": "AppsFlyer/1 CFNetwork/978.0.7 Darwin/18.6.0",
```
Comment (arnab-p) [inline __tests__/data/appsflyer_source_output.json:124]: Expected: `message.context.userAgent`

File: __tests__/data/appsflyer_source_output.json [hunk 13]
```diff
@@ -0,0 +1,141 @@
...
+      "af_ad_type": null,
+      "carrier": null,
+      "event_name": "My Apps",
+      "af_sub_siteid": null,
+      "advertising_id": null,
+      "os_version": "12.3.1",
+      "platform": "ios",
+      "af_sub3": null,
+      "contributor_3_match_type": null,
+      "selected_timezone": "UTC",
```
Comment (arnab-p) [inline __tests__/data/appsflyer_source_output.json:121]: Expected: `message.context.timezone`

File: __tests__/data/appsflyer_source_output.json [hunk 14]
```diff
@@ -0,0 +1,141 @@
...
+      "app_id": "id1217828636",
+      "contributor_1_touch_time": null,
+      "event_revenue": null,
+      "af_ad_type": null,
+      "carrier": null,
+      "event_name": "My Apps",
+      "af_sub_siteid": null,
+      "advertising_id": null,
+      "os_version": "12.3.1",
+      "platform": "ios",
```
Comment (arnab-p) [inline __tests__/data/appsflyer_source_output.json:118]: Expected: `message.platform` and in `message.context.os.name`

Also,

```
if (platform === "ios") {
  context.device.advertisingId = <value of `idfa`>;
} else {
  context.device.advertisingId = <value of `android_id`>;
}

if (context.device.advertisingId) {
  context.device.adTrackingEnabled = true;
}
```

File: __tests__/data/appsflyer_source_output.json [hunk 15]
```diff
@@ -0,0 +1,141 @@
...
+      "language": "en-US",
+      "app_id": "id1217828636",
+      "contributor_1_touch_time": null,
+      "event_revenue": null,
+      "af_ad_type": null,
+      "carrier": null,
+      "event_name": "My Apps",
+      "af_sub_siteid": null,
+      "advertising_id": null,
+      "os_version": "12.3.1",
```
Comment (arnab-p) [inline __tests__/data/appsflyer_source_output.json:117]: Expected: `message.context.os.version`

File: __tests__/data/appsflyer_source_output.json [hunk 16]
```diff
@@ -0,0 +1,141 @@
...
+      "device_download_time": "2019-01-20 04:51:16.000",
+      "contributor_1_touch_type": null,
+      "af_reengagement_window": null,
+      "af_siteid": null,
+      "language": "en-US",
+      "app_id": "id1217828636",
+      "contributor_1_touch_time": null,
+      "event_revenue": null,
+      "af_ad_type": null,
+      "carrier": null,
```
Comment (arnab-p) [inline __tests__/data/appsflyer_source_output.json:113]: Expected: `message.context.network.carrier`

File: __tests__/data/appsflyer_source_output.json [hunk 17]
```diff
@@ -0,0 +1,141 @@
...
+      "af_sub2": null,
+      "original_url": null,
+      "contributor_2_campaign": null,
+      "android_id": null,
+      "contributor_3_media_source": null,
+      "af_adset": null,
+      "af_ad": null,
+      "state": "57",
+      "network_account_id": null,
+      "device_type": "iPhoneXR",
```
Comment (arnab-p) [inline __tests__/data/appsflyer_source_output.json:97]: Expected: `message.context.device.model`

File: v0/sources/appsflyer/transform.js [hunk 18]
```diff
@@ -0,0 +1,50 @@
+const Message = require("../message");
+
+const { removeUndefinedAndNullValues } = require("../../util");
+
+function processEvent(event) {
+  const messageType = "track";
+
+  if (event.event_name) {
+    const eventName = event.event_name;
+    const message = new Message(`Appsflyer`);
```
Comment (arnab-p) [inline v0/sources/appsflyer/transform.js:10]: On the destination side, we name it as `AF`. I think we should use that here.

**Review by arnab-p — CHANGES_REQUESTED** [review]

File: __tests__/data/appsflyer_source_output.json (hunk 1 above)
Comment (prabrishac) [reply __tests__/data/appsflyer_source_output.json:19]: have put in mapping json.

File: __tests__/data/appsflyer_source_output.json (hunk 2 above)
Comment (prabrishac) [reply __tests__/data/appsflyer_source_output.json:22]: have put in mapping json.

File: __tests__/data/appsflyer_source_output.json (hunk 3 above)
Comment (prabrishac) [reply __tests__/data/appsflyer_source_output.json:29]: done

File: __tests__/data/appsflyer_source_output.json (hunk 4 above)
Comment (prabrishac) [reply __tests__/data/appsflyer_source_output.json:34]: done

File: __tests__/data/appsflyer_source_output.json (hunk 5 above)
Comment (prabrishac) [reply __tests__/data/appsflyer_source_output.json:41]: done

File: __tests__/data/appsflyer_source_output.json (hunk 6 above)
Comment (prabrishac) [reply __tests__/data/appsflyer_source_output.json:43]: done

File: __tests__/data/appsflyer_source_output.json (hunk 7 above)
Comment (prabrishac) [reply __tests__/data/appsflyer_source_output.json:44]: done

File: __tests__/data/appsflyer_source_output.json (hunk 8 above)
Comment (prabrishac) [reply __tests__/data/appsflyer_source_output.json:60]: done

File: __tests__/data/appsflyer_source_output.json (hunk 9 above)
Comment (prabrishac) [reply __tests__/data/appsflyer_source_output.json:64]: done.

File: __tests__/data/appsflyer_source_output.json (hunk 10 above)
Comment (prabrishac) [reply __tests__/data/appsflyer_source_output.json:79]: done

File: __tests__/data/appsflyer_source_output.json (hunk 11 above)
Comment (prabrishac) [reply __tests__/data/appsflyer_source_output.json:81]: have put in mapping json.

File: __tests__/data/appsflyer_source_output.json (hunk 12 above)
Comment (prabrishac) [reply __tests__/data/appsflyer_source_output.json:124]: done

File: __tests__/data/appsflyer_source_output.json (hunk 13 above)
Comment (prabrishac) [reply __tests__/data/appsflyer_source_output.json:121]: done

File: __tests__/data/appsflyer_source_output.json (hunk 14 above)
Comment (prabrishac) [reply __tests__/data/appsflyer_source_output.json:118]: done

File: __tests__/data/appsflyer_source_output.json (hunk 15 above)
Comment (prabrishac) [reply __tests__/data/appsflyer_source_output.json:117]: done

File: __tests__/data/appsflyer_source_output.json (hunk 16 above)
Comment (prabrishac) [reply __tests__/data/appsflyer_source_output.json:113]: done

File: __tests__/data/appsflyer_source_output.json (hunk 17 above)
Comment (prabrishac) [reply __tests__/data/appsflyer_source_output.json:97]: done

File: v0/sources/appsflyer/transform.js (hunk 18 above)
Comment (prabrishac) [reply v0/sources/appsflyer/transform.js:10]: done.

File: v0/sources/appsflyer/mapping.json [hunk 19]
```diff
@@ -0,0 +1,18 @@
+{
+  "customer_user_id": ["userId", "context.userId"],
```
Comment (arnab-p) [inline v0/sources/appsflyer/mapping.json:2]: It should be `context.traits.userId`

File: v0/sources/appsflyer/transform.js [hunk 20]
```diff
@@ -0,0 +1,68 @@
...
+    message.setEventName(eventName);
+
+    const properties = { ...event };
+    message.setProperty("properties", properties);
+
+    // set fields in payload from mapping json
+    message.setProperties(event, mappingJson);
+
+    // Remove the fields from properties that are already mapped to other fields.
+    Object.keys(mappingJson).forEach(key => {
```
Comment (arnab-p) [inline v0/sources/appsflyer/transform.js:29]: Removing the mapped properties earlier, it's not handling the keys `idfa`, `appsflyer_id` etc.

File: v0/sources/appsflyer/transform.js [hunk 21]
```diff
@@ -0,0 +1,68 @@
...
+    return null;
+  }
+  throw new Error("Unknwon event type from Appsflyer");
+}
+
+function process(event) {
+  let returnValue = {};
+  const response = processEvent(event);
+  returnValue = removeUndefinedAndNullValues(response);
+  return returnValue;
```
Comment (arnab-p) [inline v0/sources/appsflyer/transform.js:65]:
```suggestion
  return removeUndefinedAndNullValues(processEvent(event));
```

Should we handle inline?

**Review by arnab-p — CHANGES_REQUESTED** [review]

**Review by arnab-p — APPROVED** [review]

**Review by arnab-p — APPROVED** [review]

**Pull request merged**